File: java/src/main/java/ml/wonwoo/springdatashowcase/product/Catalog.java
package ml.wonwoo.springdatashowcase.product;

import ml.wonwoo.springdatashowcase.support.Identifier;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.NoRepositoryBean;
import org.springframework.data.util.Streamable;

@NoRepositoryBean
public interface Catalog<T extends Product> extends JpaRepository<T, Identifier> {

    Streamable<T> findByName(String name);

    @Query("select p from #{#entityName} p where p.name = :name")
    Streamable<T> findName(String name);
}

File: kotlin/src/main/kotlin/ml/wonwoo/springdatashowcase/product/Catalog.kt
package ml.wonwoo.springdatashowcase.product

import ml.wonwoo.springdatashowcase.support.Identifier
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.NoRepositoryBean

@NoRepositoryBean
interface Catalog<T : Product> : JpaRepository<T, Identifier> {

    fun findByName(name: String): List<T>

    @Query("select p from #{#entityName} p where p.name = :name")
    fun findName(name: String): List<T>

}